src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -476,6 +476,19 @@ private void AdvanceReader(BufferSegment consumedSegment, int consumedIndex, Buf
             TrySchedule(_writerScheduler, completionData);
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private long GetLength(BufferSegment startSegment, int startIndex, BufferSegment endSegment, int endIndex)
+        {
+            if (startSegment != endSegment)
+            {
+                // (End offset) - (start offset)
+                return (endSegment.RunningIndex + endIndex) - (startSegment.RunningIndex + startIndex);
```

Comment:
When benchmarking, it might be worth testing just returning this expression instead of branching.